src/proof/lfsc/lfsc_node_converter.cpp

```diff
@@ -0,0 +1,1013 @@
+      TypeNode selt = nm->mkFunctionType({d_sortType, intType}, fselt);
+      Node sel = getSymbolInternal(k.getKind(), selt, "sel");
+      Node kn = typeAsNode(convertType(tn.getSelectorRangeType()));
+      Assert(!cacheVal.isNull() && cacheVal.getKind() == CONST_RATIONAL);
+      return nm->mkNode(APPLY_UF, sel, kn, cacheVal);
+    }
+    else if (sfi == SkolemFunId::RE_UNFOLD_POS_COMPONENT)
+    {
+      TypeNode strType = nm->stringType();
+      TypeNode reType = nm->regExpType();
+      TypeNode intType = nm->integerType();
+      TypeNode reut = nm->mkFunctionType({strType, reType, intType}, strType);
+      Node sk = getSymbolInternal(k.getKind(), reut, "skolem_re_unfold_pos");
+      Assert(!cacheVal.isNull() && cacheVal.getKind() == SEXPR
+             && cacheVal.getNumChildren() == 3);
+      // third value is mpz, which is not converted
+      return nm->mkNode(
+          APPLY_UF,
+          {sk, convert(cacheVal[0]), convert(cacheVal[1]), cacheVal[2]});
+    }
+  }
+  return Node::null();
+}
+
+Node LfscNodeConverter::typeAsNode(TypeNode tni) const
+{
+  // should always exist in the cache, as we always run types through
+  // postConvertType before calling this method.
+  std::map<TypeNode, Node>::const_iterator it = d_typeAsNode.find(tni);
+  AlwaysAssert(it != d_typeAsNode.end()) << "Missing typeAsNode " << tni;
+  return it->second;
+}
+
+Node LfscNodeConverter::mkInternalSymbol(const std::string& name, TypeNode tn)
+{
+  Node sym = NodeManager::currentNM()->mkBoundVar(name, tn);
+  d_symbols.insert(sym);
+  return sym;
+}
+
+Node LfscNodeConverter::getSymbolInternalFor(Node n, const std::string& name)
+{
+  return getSymbolInternal(n.getKind(), n.getType(), name);
+}
+
+Node LfscNodeConverter::getSymbolInternal(Kind k,
+                                          TypeNode tn,
+                                          const std::string& name)
+{
+  std::tuple<Kind, TypeNode, std::string> key(k, tn, name);
+  std::map<std::tuple<Kind, TypeNode, std::string>, Node>::iterator it =
+      d_symbolsMap.find(key);
+  if (it != d_symbolsMap.end())
+  {
+    return it->second;
+  }
+  Node sym = mkInternalSymbol(name, tn);
+  d_symbolToBuiltinKind[sym] = k;
+  d_symbolsMap[key] = sym;
+  return sym;
+}
+
+void LfscNodeConverter::getCharVectorInternal(Node c, std::vector<Node>& chars)
+{
+  Assert(c.getKind() == CONST_STRING);
+  NodeManager* nm = NodeManager::currentNM();
+  const std::vector<unsigned>& vec = c.getConst<String>().getVec();
+  if (vec.size() == 0)
+  {
+    Node ec = getSymbolInternalFor(c, "emptystr");
+    chars.push_back(ec);
+    return;
+  }
+  TypeNode tnc = nm->mkFunctionType(nm->integerType(), c.getType());
+  Node aconstf = getSymbolInternal(CONST_STRING, tnc, "char");
+  for (unsigned i = 0, size = vec.size(); i < size; i++)
+  {
+    Node cc = nm->mkNode(APPLY_UF, aconstf, nm->mkConst(Rational(vec[i])));
+    chars.push_back(cc);
+  }
+}
+
+bool LfscNodeConverter::isIndexedOperatorKind(Kind k)
```

Comment:
Maybe this can reside in a more general location? Looks like a generally useful utility.

Reply:
Thinking about this some more, I don't think we want this as a general utility.

I'll clarify that this method is "is k a kind that is printed as an indexed operator in the LFSC format".